sys/psa_crypto/psa_crypto_algorithm_dispatch.c

```diff
@@ -542,19 +543,150 @@ psa_status_t psa_algorithm_dispatch_import_key(const psa_key_attributes_t *attri
     }
     return psa_builtin_import_key(attributes, data, data_length, key_data, key_data_size, key_bytes, bits);
 }
+
 #endif /* MODULE_PSA_KEY_MANAGEMENT */
 
 #if IS_USED(MODULE_PSA_CIPHER)
-psa_status_t psa_algorithm_dispatch_cipher_encrypt( const psa_key_attributes_t *attributes,
-                                                    psa_algorithm_t alg,
-                                                    const psa_key_slot_t *slot,
+
+psa_status_t psa_algorithm_dispatch_cipher_encrypt_setup(   psa_cipher_operation_t *operation,
+                                                            const psa_key_attributes_t *attributes,
+                                                            const psa_key_slot_t *slot,
+                                                            psa_algorithm_t alg)
+{
+    operation->alg = PSA_ENCODE_CIPHER_OPERATION(alg, attributes->type, attributes->bits);
+
+    uint8_t *key_data = NULL;
+    size_t *key_bytes = NULL;
+
+    psa_get_key_data_from_key_slot(slot, &key_data, &key_bytes);
+
+    if (operation->alg == PSA_INVALID_OPERATION) {
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+
+    switch (operation->alg)
+    {
+    #if IS_USED(MODULE_PSA_CIPHER_CHACHA20)
+        case PSA_STREAM_CIPHER_CHACHA20:
+            return psa_cipher_chacha20_encrypt_setup(&operation->backend_ctx.cipher_ctx.chacha20, key_data, *key_bytes);
+    #endif
+    default:
+        (void)operation;
+        (void)attributes;
+        (void)slot;
+        (void)alg;
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+}
+
+psa_status_t psa_algorithm_dispatch_cipher_decrypt_setup(   psa_cipher_operation_t *operation,
+                                                            const psa_key_attributes_t *attributes,
+                                                            const psa_key_slot_t *slot,
+                                                            psa_algorithm_t alg)
+{
+    operation->alg = PSA_ENCODE_CIPHER_OPERATION(alg, attributes->type, attributes->bits);
+
+    uint8_t *key_data = NULL;
+    size_t *key_bytes = NULL;
+
+    psa_get_key_data_from_key_slot(slot, &key_data, &key_bytes);
+
+    if (operation->alg == PSA_INVALID_OPERATION) {
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
```

Comment:
Can `operation->alg` really be `PSA_INVALID_OPERATION`?

Reply:
I see, so maybe the member being called `alg` was confusing to me, as I would have expected it to be of type `psa_algorithm_t` and not our custom enum type. Do you have an idea for a better name?

Reply:
maybe name it cipher_alg to distinguish?

Reply:
Or something like `cipher_instance`? Also, adding a comment to that member of the operation struct would be nice (if not present already).

Reply:
`cipher_instance` sounds good? comment will also be done